morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,378 @@
+                          nlohmann::json& x_headers)
+{
+    method = query.value("method", "GET");
+    std::transform(method.begin(), method.end(), method.begin(), ::toupper);
+
+    endpoint = query.value("endpoint", "");
+    if (endpoint.empty())
+    {
+        throw std::runtime_error("'REST loader' receives query with empty endpoint");
+    }
+
+    content_type = query.value("content_type", "");
+    body         = query.value("body", "");
+    params       = query.value("params", nlohmann::json());
+    x_headers    = query.value("x-headers", nlohmann::json());
+}
+
+http::verb get_http_verb(const std::string& method)
+{
+    http::verb verb;
+    if (method == "GET")
+    {
+        verb = http::verb::get;
+    }
+    else if (method == "POST")
+    {
+        verb = http::verb::post;
+    }
+    else
+    {
+        throw std::runtime_error("'REST Loader' receives method not supported: " + method);
+    }
+    return verb;
+}
+
+void construct_request(http::request<http::string_body>& request,
+                       const std::string& method,
+                       const std::string& host,
+                       const std::string& target,
+                       const std::string& query,
+                       const std::string& content_type,
+                       const std::string& body,
+                       const nlohmann::json& x_headers)
+{
+    auto verb = get_http_verb(method);
+    request.method(verb);
+    request.target(target + query);
+    request.version(11);
+    request.set(http::field::host, host);
+    request.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+    if (verb == http::verb::post)
+    {
+        request.set(http::field::content_type, content_type);
+        request.body() = body;
+        request.prepare_payload();
+    }
+    for (auto& x_header : x_headers.items())
+    {
+        request.insert(x_header.key(), x_header.value());
+    }
+}
+
+void get_response_with_retry(http::request<http::string_body>& request,
+                             http::response<http::dynamic_body>& response,
+                             int max_retry,
+                             int retry_interval_milliseconds)
+{
+    net::io_context ioc;
+    tcp::resolver resolver(ioc);
+    beast::tcp_stream stream(ioc);
+    auto const endpoint_results = resolver.resolve(std::string(request[http::field::host]), PORT);
+
+    while (max_retry > 0)
+    {
+        stream.connect(endpoint_results);
+        http::write(stream, request);
+        beast::flat_buffer buffer;
+        http::read(stream, buffer, response);
+        beast::error_code ec;
+        stream.socket().shutdown(tcp::socket::shutdown_both, ec);
+        if (ec && ec != beast::errc::not_connected)
+        {
+            throw beast::system_error{ec};
+        }
+        int status = response.result_int();
+        // Retry only if status is 503 Service Unavailable / 504 Gateway Timeout
+        if (status != 503 && status != 504)
```

Comment:
Using named constants for HTTP status codes `503` and `504` would be a good for reusability.